### PR TITLE
reduce the overhead of BType#toString

### PR DESCRIPTION
**Pull request by mkeskells**
avoid allocation of StringBuilder, char[] and String for trivial cases

### CODE REVIEW

**Comment by mkeskells** [comment]
cc: @hrhino @retronym

File: src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala [hunk 1]
```diff
@@ -268,7 +260,7 @@ abstract class BTypes {
     def asPrimitiveBType : PrimitiveBType = this.asInstanceOf[PrimitiveBType]
   }
 
-  sealed trait PrimitiveBType extends BType {
+  sealed class PrimitiveBType(override val toString: String) extends BType {
```
Comment (hrhino) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:263]: Could this be abstract?

Or it could be final and the instances vals

**Review by hrhino — APPROVED** [review]

**Comment by retronym** [comment]
I remember reviewing this before and raising questions over whether retaining these vals for the duration of the `jvm` phase could have a negative effect of memory (more work to do in old-gen collections if these `String`-s make it there and higher retained memory) that might counteract the gain of reducing allocation of short lived strings (reduced frequency of old-gen collections).

**Comment by retronym** [comment]
Oh wait, this is just for the primitive `BTypes`, not for all of them. I've made part of this change already, albeit only on 2.13.x. Let's just backport #7821 and then add in the `PrimitiveBType.toString` val.

**Comment by retronym** [comment]
I've done the backport and reapplication of this patch in #8457

File: src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala (hunk 1 above)
Comment (retronym) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:263]: Looks like its a sealed class for pattern matching over the cases. We don't have a way to do the same with val instances (although one could switch to Java enums for the same effect.)

Comment (mkeskells) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:263]: made abstract to make the matches exhaustive
@retronym WDYT about making the class final and making the matched properties in the rest of the class based on constructor params?

Comment (retronym) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:263]: It's use in patterns is pretty widespread so I would only do this if there is a some good motivating reason for the change.

File: src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala [hunk 2]
```diff
@@ -618,6 +610,7 @@ abstract class BTypes {
    * infos using `get`, but it reports inliner warnings for missing infos that prevent inlining.
    */
   sealed abstract class ClassBType protected(val internalName: InternalName) extends RefBType {
+    override val toString = "L" + internalName+';'
```
Comment (mkeskells) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:613]: cache the toString

**Pull request closed without merging**